include/seqan3/index/fm_index.hpp

```diff
@@ -0,0 +1,345 @@
+protected:
+    //!\privatesection
+
+    //!\brief The type of the underlying SDSL index.
+    using sdsl_index_type = typename fm_index_traits::sdsl_index_type;
+    /*!\brief The type of the reduced alphabet type. (The reduced alphabet might be smaller than the original alphabet
+     *        in case not all possible characters occur in the indexed text.)
+     */
+    using comp_char_type = typename sdsl_index_type::alphabet_type::comp_char_type;
+
+    //!\brief Underlying index from the SDSL.
+    sdsl_index_type m_index;
+    //!\brief Pointer to the indexed text.
+    text_t const * text = nullptr;
+
+    //!\publicsection
+
+public:
+    //!\brief The type of the indexed text.
+    using text_type = text_t;
+    //!\brief The type of the underlying character of text_type.
+    using char_type = innermost_value_type_t<text_t>;
+    //!\brief Type for representing positions in the indexed text.
+    using size_type = typename sdsl_index_type::size_type;
+
+    //!\brief The type of the (unidirectional) iterator.
+    using iterator_type = fm_index_iterator<fm_index<text_t, fm_index_traits>>;
+    friend class fm_index_iterator<fm_index<text_t, fm_index_traits>>;
+    friend class detail::fm_index_iterator_node<fm_index<text_t, fm_index_traits>>;
+
+    // static constexpr bool is_bidirectional = false;
+    // static constexpr uint8_t dimensions = DIMENSIONS;
+
+    /*!\name Constructors and destructor
+     * \{
+     */
+    fm_index() = default;
+    fm_index(fm_index const &) = default;
+    fm_index & operator=(fm_index const &) = default;
+    fm_index(fm_index &&) = default;
+    fm_index & operator=(fm_index &&) = default;
+    ~fm_index() = default;
+    //!\}
+
+    /*!\brief Constructor that immediately constructs the index given a range.
+     *        The range cannot be an rvalue (i.e. a temporary object).
+     * \tparam text_t The type of range to construct from; must satisfy seqan3::forward_range_concept.
+     * \param[in] text The text to construct from.
+     *
+     * ### Complexity
+     *
+     * \todo At least linear.
+     *
+     * ### Exceptions
+     *
+     * No guarantees.
+     */
+    fm_index(text_t const & text)
+    {
+        construct(text);
+    }
+
+    //!\overload
+    fm_index(text_t &&) = delete;
+
+    //!\overload
+    fm_index(text_t const &&) = delete;
+
+    /*!\brief Constructs the index given a range. The range cannot be an rvalue (i.e. a temporary object).
+     *        \todo Poorely implemented with regard to the memory peak due to not matching interfaces with the SDSL
+     * \tparam text_t The type of range to construct from; must satisfy seqan3::forward_range_concept.
+     * \param[in] text The text to construct from.
+     *
+     * ### Complexity
+     *
+     * \todo At least linear.
+     *
+     * ### Exceptions
+     *
+     * No guarantees.
+     */
+    void construct(text_t const & text)
+    {
+        assert(!text.empty());
+        this->text = &text;
+        // TODO:
```

Comment:
Should also be added to doxygen todos

Reply:
I wouldn't include implementation TODOs into the doxygen TODO-list.

I'll create some issues on Github later